### ISSUE TITLE
File input w/ multiline codec merges lines from different files.

### ISSUE THREAD

**Issue by jordansissel** [elastic/logstash#1754]
Reproducing:

```
% cat > /tmp/x.a
Hello

% cat > /tmp/x.b
   world
  fizzle

% cat > /tmp/x.a
fancy pants
```

Logstash config:
```
% bin/logstash -e 'input { file { path => "/tmp/x.?" codec => multiline { pattern => "^\s" what => previous } } }'
{
  "@timestamp" => "2014-09-19T00:12:14.243Z",
  "message"    => "Hello\n   world\n  fizzle",
  "@version"   => "1",
  "tags"       => [
    [0] "multiline"
  ],
  "host"       => "oh-my",
  "path"       => "/tmp/x.a"
}
```

Unexpectedly, the lines from /tmp/x.b are included in a message from /tmp/x.a.

A brief review of the file input makes the cause of this pretty clear; the @codec is not unique per file.

**Comment by codearoo** [comment]
Hi...  wouldn't you want to not use multiline in codec, but rather in filter? If you use it there, it has the MUCH NEEDED stream_identity property which you can set to something like "%{host}.%{FileName}" if you generate the file name in however it gets input?

**Comment by jordansissel** [comment]
@codearoo the bug described in this ticket is that the file input uses the same codec for all files, when it should internally have separate streams for each file. Your note on stream_identity is exactly the bug being described :)

**Comment by dsolsona** [comment]
Also the problem with using the multiline with filter is that you can't then use multiple filter workers because it's not thread safe (https://github.com/elasticsearch/logstash/issues/1590)

**Comment by miguno** [comment]
@jordansissel Is there a workaround for this bug?  In our testing both multiline codec and multiline filter suffer from this problem.

**Comment by varh1i** [comment]
Could anyone suggest a solution for this?

- we would use multiple worker threads (for performance) -> multiline filter doesn't work (https://github.com/elasticsearch/logstash/issues/1590, https://github.com/elasticsearch/logstash/issues/1591)
- multiline codec (in lumberjack input) works but we encounter the problem described in this ticket. 

solutions for now:
- using multiple logstash-forwarder and send them to different lumberjack port (scales very poorly: new logstash-forwarder for each logfile that has multilines)
- using one logstash-receiver with redis output and new logstash indexer that reads from redis but this way we need to add redis and a separate logstash reciever that deals with it.(https://groups.google.com/forum/#!msg/logstash-users/fUhulxmZ0Ek/HfGev7BOhCUJ)

is there any better solution?

**Comment by jmreicha** [comment]
Experiencing this as well on 1.5.0 RC2.

**Comment by celesteking** [comment]
So, not only filter, but multiline codec isn't thread-safe? This should really be noted in docs, as it's very misleading. We're going to deploy serious multi-connection multi-line logging and having such undocumented obstacles is inadmissible. 

https://logstash.jira.com/browse/LOGSTASH-159 didn't fix the filter, I guess.

**Comment by yeroc** [comment]
Any update on this? It seems like the only workaround is to keep separate lumberjack inputs on different ports but this isn't scalable!  Is there another workaround or imminent fix?

**Comment by dsolsona** [comment]
The only workaround I found was replacing logstash-forwarder with log-courier which has multiline support on the client side.

**Comment by yeroc** [comment]
Thanks @dsolsona, log-courier does appear to be more actively maintained.  Will take a look once logstash 1.5 is released which promises to make plugin installation easier.

**Comment by ph** [comment]
@yeroc the current version of the lumberjack input create a codec instance per connection, so he will be able to merge events from a same source.

**Comment by yeroc** [comment]
@ph My experience with the current production version is that it doesn't even handle that case.  With the multiline codec I'm seeing events merged from two different source machines (same port but different connections).

**Comment by ph** [comment]
@yeroc np you were right concerning 1.4.2 :)

**Comment by jordansissel** [comment]
For Logstash 1.5.0, we've moved all plugins to individual repositories, so I have moved this issue to https://github.com/logstash-plugins/logstash-input-file/issues/44. Let's continue the discussion there! :)

**Comment by rajeshdommati** [comment]
is this issue fixed in any of the  Logstash 1.5.* releases.
